### PR TITLE
Make istio object in Domain yaml work

### PR DESCRIPTION
**Pull request by davidzl**

### CODE REVIEW

**Comment by davidzl** [comment]
Ryan,

Please merge my latest commit (https://github.com/oracle/weblogic-kubernetes-operator/pull/1160/commits/d2ca72b2c0cbbb2dc0b5535ea3e13016e20f3f1d) to make it work.

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml [hunk 1]
```diff
@@ -136,4 +136,7 @@ createDomainScriptName: create-domain-job.sh
 createDomainFilesDir: wlst
 
 # istioEnabled specifies whether or not it is deployed under Istio.
-istioEnabled: false
\ No newline at end of file
+istioEnabled: true
```
Comment (markxnelson) [inline kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml]: istio should not be enabled by default in the sample

File: model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java [hunk 2]
```diff
@@ -11,25 +11,47 @@
 
 public class Istio {
 
-  @Description("True, if this domain is deployed under an Istio service mesh. Defaults to true. Not required.")
-  private Boolean istioEnabled = true;
+  @Description(
+      "True, if this domain is deployed under an Istio service mesh. Defaults to true. Not required.")
```
Comment (markxnelson) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java]: istio should default to false

File: model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java [hunk 3]
```diff
@@ -11,25 +11,47 @@
...
    *
-   * @param istioEnabled True, if this domain is deployed under an Istio service mesh.
+   * @param enabled True, if this domain is deployed under an Istio service mesh.
+   */
+  public void setEnabled(Boolean enabled) {
+    this.enabled = enabled;
+  }
+
+  /**
+   * Get the readyness port.
```
Comment (markxnelson) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java]: spelling error

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/create-domain-script.sh [hunk 4]
```diff
@@ -10,4 +10,6 @@ export DOMAIN_HOME=${DOMAIN_HOME_DIR}
 # Create the domain
 wlst.sh -skipWLSModuleScanning ${CREATE_DOMAIN_SCRIPT_DIR}/create-domain.py
 
-
+if [ "${ISTIO_ENABLED}" == "true" ]; then
+  wlst.sh -skipWLSModuleScanning ${CREATE_DOMAIN_SCRIPT_DIR}/istio-fix-domain.py
```
Comment (markxnelson) [inline kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/create-domain-script.sh]: seems like there should be some error checking in here?

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
+# Copyright 2018, 2019, Oracle Corporation and/or its affiliates.  All rights reserved.
```
Comment (markxnelson) [inline kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py]: "2018" should not be in the copyright notice - this is a new file right? so just "2019"

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+t3_public_address            = getEnvVar("T3_PUBLIC_ADDRESS")
+number_of_ms                 = int(getEnvVar("CONFIGURED_MANAGED_SERVER_COUNT"))
+cluster_type                 = getEnvVar("CLUSTER_TYPE")
+managed_server_name_base     = getEnvVar("MANAGED_SERVER_NAME_BASE")
+managed_server_name_base_svc = getEnvVar("MANAGED_SERVER_NAME_BASE_SVC")
+domain_logs                  = getEnvVar("DOMAIN_LOGS_DIR")
+script_dir                   = getEnvVar("CREATE_DOMAIN_SCRIPT_DIR")
+istio_readiness_port         = int(getEnvVar("ISTIO_READINESS_PORT"))
+
+# Update Domain for Istio
```
Comment (markxnelson) [inline kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py]: i would like to see some logging in this file, print out what is going on, so it is easier for us to debug it of we need to

File: model/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java [hunk 7]
```diff
@@ -486,6 +505,20 @@ public void setConfigOverrideSecrets(@Nullable List<String> overridesSecretNames
     this.configOverrideSecrets = overridesSecretNames;
   }
 
+  boolean istioEnabled() {
```
Comment (markxnelson) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java]: javadoc missing

File: operator/src/main/java/oracle/kubernetes/operator/helpers/CrdHelper.java [hunk 8]
```diff
@@ -114,8 +111,7 @@ static V1beta1CustomResourceDefinition createModel(KubernetesVersion version) {
 
     static V1ObjectMeta createMetadata() {
       return new V1ObjectMeta()
-          .name(KubernetesConstants.CRD_NAME)
-          .putLabelsItem(LabelConstants.RESOURCE_VERSION_LABEL, DEFAULT_OPERATOR_VERSION);
```
Comment (markxnelson) [inline operator/src/main/java/oracle/kubernetes/operator/helpers/CrdHelper.java]: why are these labels removed?

File: operator/src/main/java/oracle/kubernetes/operator/helpers/JobStepContext.java [hunk 9]
```diff
@@ -196,6 +200,7 @@ protected V1JobSpec createJobSpec(TuningParameters tuningParameters) {
 
   private V1PodTemplateSpec createPodTemplateSpec(TuningParameters tuningParameters) {
     V1ObjectMeta metadata = new V1ObjectMeta().name(getJobName());
+    if (istioEnabled()) metadata.putAnnotationsItem("sidecar.istio.io/inject", "false");
```
Comment (markxnelson) [inline operator/src/main/java/oracle/kubernetes/operator/helpers/JobStepContext.java]: if statements need braces

File: operator/src/main/resources/scripts/introspectDomain.py [hunk 10]
```diff
@@ -853,7 +853,7 @@ def customizeNetworkAccessPoint(self, nap, listen_address):
     # FWIW there's theoretically no need to 'add' or 'replace' when empty
     #   since the runtime default is the server listen-address.
     nap_name=nap.getName()
-    if not (nap.getListenAddress() is None) and len(nap.getListenAddress()) > 0:
+    if not (nap.getListenAddress() is None) and len(nap.getListenAddress()) > 0 and not (nap_name.startswith('istio-')) and not (nap_name == 'T3Channel'):
```
Comment (markxnelson) [inline operator/src/main/resources/scripts/introspectDomain.py]: check the logic here - does this break the non-istio case?

**Comment by davidzl** [comment]
Added 2 more commits to address Mark's comments.

File: operator/src/main/resources/scripts/startNodeManager.sh [hunk 11]
```diff
@@ -255,12 +255,14 @@ EOF
     if [ "${ADMIN_PORT_SECURE}" = "true" ]; then
       admin_protocol="https"
     fi  
-    echo "AdminURL=$admin_protocol\\://${AS_SERVICE_NAME}\\:${ADMIN_PORT}" >> ${wl_props_file}
+    if [ "${ISTIO_ENABLED}" == "true" ]; then
+      echo "AdminURL=t3\\://${AS_SERVICE_NAME}\\:${ADMIN_PORT}" >> ${wl_props_file}
```
Comment (tbarnes-us) [inline operator/src/main/resources/scripts/startNodeManager.sh]: Why is AdminURL forced to t3 here?

Comment (tbarnes-us) [reply operator/src/main/resources/scripts/startNodeManager.sh]: If it needs to be t3 for some reason, then t3s should be used if the admin port is secured.

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-job-template.yaml [hunk 12]
```diff
@@ -67,6 +71,10 @@ spec:
               value: "%DOMAIN_HOME%"
             - name: DOMAIN_LOGS_DIR
               value: "%DOMAIN_ROOT_DIR%/logs/%DOMAIN_UID%"
+            - name: ISTIO_ENABLED
+              value: "%ISTIO_ENABLED%"
+            - name: ISTIO_READINESS_PORT
+              value: "%ISTIO_READINESS_PORT%"
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-job-template.yaml]: We shouldn't be requiring users to set these values - the Operator itself should set them for both the introspector job and the wl pods when istio is enabled via the domain yaml. If we want to keep it here, maybe rename it ISTIO_ENABLED_EXPERIMENTAL or at least put a clear comment above the env var stating that it's experimental.

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+readDomain(domain_path)
+
+cd('/Servers/' + admin_server_name)
+create('istio-probe', 'NetworkAccessPoint')
+cd('/Servers/%s/NetworkAccessPoints/istio-probe' % admin_server_name)
+set('Protocol', 'http')
+set('ListenAddress', '127.0.0.1')
+set('PublicAddress', domain_uid + '-' + admin_server_name)
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py]: (1) For all addresses, the name needs to be converted to be DNS legal - see introspectDomain.py for example usage:
```
def toDNS1123Legal(self, address):
    return address.lower().replace('_','-')
```
(2) Ideally we should not be requiring that users set addresses in the original configuration. Instead, we should be setting the ListenAddress and Public Address of istio enabled channels for the user via situational configuration generated in 'introspectDomain.py' instead of requiring customers to set them exactly correctly.
(3) Note that the public-address of a T3Channel needs to be appropriate for the client's location, so "domain_uid + '-' + admin_server_name" can be wrong if the client is external. The sample should somehow note this (I assume most customers will use config-overrides to set the value).

File: kubernetes/charts/weblogic-operator/templates/_operator-dep.tpl [hunk 14]
```diff
@@ -36,6 +36,8 @@ spec:
           value: "false"
         - name: "JAVA_LOGGING_LEVEL"
           value: {{ .javaLoggingLevel | quote }}
+        - name: ISTIO_ENABLED
+          value: {{ .istioEnabled | quote }}
```
Comment (tbarnes-us) [inline kubernetes/charts/weblogic-operator/templates/_operator-dep.tpl]: Why is this needed?  Can the value instead be derived from the domain resource as needed?

File: kubernetes/charts/weblogic-operator/values.yaml [hunk 15]
```diff
@@ -96,3 +96,6 @@ elasticSearchHost: "elasticsearch.default.svc.cluster.local"
 # elasticSearchPort specifies the port number of where elasticsearch is running.
 # This parameter is ignored if 'elkIntegrationEnabled' is false.
 elasticSearchPort: 9200
+
+# istioEnabled specifies whether or not it is deployed under Istio.
+istioEnabled: false
```
Comment (tbarnes-us) [inline kubernetes/charts/weblogic-operator/values.yaml]: Maybe make clear that this is an experimental setting (mark it with '(experimental)')

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml [hunk 16]
```diff
@@ -133,3 +134,9 @@ createDomainScriptName: create-domain-job.sh
 # Kubernetes config map, which in turn is mounted to the `createDomainScriptsMountPath`,
 # so that the Kubernetes pod can use the scripts and supporting files to create a domain home.
 createDomainFilesDir: wlst
+
+# istioEnabled specifies whether or not it is deployed under Istio.
+istioEnabled: false
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml]: Mark this with '(experimental)'?

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml [hunk 17]
```diff
@@ -133,3 +134,9 @@ createDomainScriptName: create-domain-job.sh
...
+
+# istioReadinessPort specifies the WebLogic port for readiness probe which is used under Istio.
+istioReadinessPort: 8888
```
Comment (tbarnes-us) [inline kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml]: Mark this with '(experimental)'?

File: operator/src/main/resources/scripts/startNodeManager.sh (hunk 11 above)
Comment (tbarnes-us) [reply operator/src/main/resources/scripts/startNodeManager.sh]: How is ISTIO_ENABLED getting passed to the pod scripts (which in turn make use of AdminURL)?  I don't see it getting set by the Operator, and I don't see the sample's domain resource template specifying the value.   The former is preferable.  For an example of the Operator setting env vars for the introspector job and for wl pods, search for "LOG_HOME" in the Operator java source.

Comment (davidzl) [reply operator/src/main/resources/scripts/startNodeManager.sh]: In the first phase of Istio support, let's support just plain ports now. Will research the secure ports in the future as separate tasks.

ISTIO_ENABLED is passed via "kubernetes/samples/scripts/common/utility.sh"

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py (hunk 13 above)
Comment (davidzl) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py]: I suspect public-address will not work under Istio. In Istio, it will implicitly forward payloads to pod name to 127.0.0.1. Not sure how it will deal with public address. I think this can be investigated in the next stage.

File: kubernetes/charts/weblogic-operator/templates/_operator-dep.tpl (hunk 14 above)
Comment (davidzl) [reply kubernetes/charts/weblogic-operator/templates/_operator-dep.tpl]: This is required in the create-domain step when the domain resource has not been generated.

File: kubernetes/charts/weblogic-operator/values.yaml (hunk 15 above)
Comment (davidzl) [reply kubernetes/charts/weblogic-operator/values.yaml]: ok

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml (hunk 16 above)
Comment (davidzl) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml]: ok

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml (hunk 17 above)
Comment (davidzl) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml]: ok

File: operator/src/main/java/oracle/kubernetes/operator/helpers/CrdHelper.java (hunk 8 above)
Comment (davidzl) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/CrdHelper.java]: that might be branch/merge issue.

File: model/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java (hunk 7 above)
Comment (davidzl) [reply model/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java]: added

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py (hunk 6 above)
Comment (davidzl) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py]: added

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py (hunk 5 above)
Comment (davidzl) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py]: fixed

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/create-domain-script.sh (hunk 4 above)
Comment (davidzl) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/create-domain-script.sh]: The wlst exception will be thrown to the pod logs if there is any.

File: model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java (hunk 3 above)
Comment (davidzl) [reply model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java]: fixed

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml (hunk 1 above)
Comment (davidzl) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/create-domain-inputs.yaml]: fixed.

File: model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java (hunk 2 above)
Comment (davidzl) [reply model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java]: fixed

File: operator/src/main/java/oracle/kubernetes/operator/helpers/JobStepContext.java (hunk 9 above)
Comment (davidzl) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/JobStepContext.java]: this is forced by the code style checking of mvn plugin.

File: operator/src/main/resources/scripts/introspectDomain.py (hunk 10 above)
Comment (davidzl) [reply operator/src/main/resources/scripts/introspectDomain.py]: fixed

File: operator/src/main/resources/scripts/startNodeManager.sh (hunk 11 above)
Comment (tbarnes-us) [reply operator/src/main/resources/scripts/startNodeManager.sh]: > ISTIO_ENABLED is passed via "kubernetes/samples/scripts/common/utility.sh"

Instead, it should be set by the operator for all jobs/pods when the operator detects the domain is istio-enabled.  Follow 'LOG_HOME' in the java source as an example...

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py (hunk 13 above)
Comment (tbarnes-us) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py]: The 'toDNS1123Legal' is still required regardless.

File: kubernetes/charts/weblogic-operator/templates/_operator-dep.tpl (hunk 14 above)
Comment (tbarnes-us) [reply kubernetes/charts/weblogic-operator/templates/_operator-dep.tpl]: Will this have a negative impact on domains that don't set the experimental istio flags?

**Comment by markxnelson** [comment]
Tiburon team are trying out this branch and will provide detailed feedback. 
@mgianatagh @markxnelson

**Comment by tbarnes-us** [comment]
> Tiburon team are trying out this branch and will provide detailed feedback.
> @mgianatagh @markxnelson

Note that specialized testing will eventually needed for paths that particularly depend on 'public addresses' that are being directly set in this branch's sample WL config.   In particular RMI smart stub reconnects and recovering interrupted multi-srerver JTA transactions.  If I understand David correctly, this testing has not been completed and he anticipates there may be problems in this area.

**Comment by markxnelson** [comment]
Got it - and we are not assuming this is "done done"

File: kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py (hunk 13 above)
Comment (davidzl) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py]: fixed.

Comment (davidzl) [reply kubernetes/samples/scripts/create-weblogic-domain/domain-home-on-pv/wlst/istio-fix-domain.py]: I assume toDNS1123Legal is also applied to the pod names, right? Otherwise, the pod name and wls public address does not match, that will be a problem.

File: kubernetes/charts/weblogic-operator/templates/_operator-dep.tpl (hunk 14 above)
Comment (davidzl) [reply kubernetes/charts/weblogic-operator/templates/_operator-dep.tpl]: It is harmless. The property istioEnabled is defined in "kubernetes/charts/weblogic-operator/values.yaml" and its value is false by default.

File: model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java (hunk 2 above)
Comment (rjeberhard) [reply model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java]: My intention was that istioEnabled here defaults to true...  The idea was that Istio is disabled by default, but enabled by default once the "istio: {}" element is included in domain.yaml.  So...

1. Default
Istio enablement is disabled

2. Istio element present
spec:
  istio: {}
Istio enablement is enabled

3. Istio element present, but disabled
spec:
  istio:
    istioEnabled = false
    ... other settings preserved
Istio enablement is disabled.

Also, I think I would rename "istioEnabled" to "enabled"

Comment (rjeberhard) [reply model/src/main/java/oracle/kubernetes/weblogic/domain/model/Istio.java]: Spacing got lost after I posted the comment...  If it's confusing, let me know.

File: operator/src/main/java/oracle/kubernetes/operator/helpers/CrdHelper.java (hunk 8 above)
Comment (rjeberhard) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/CrdHelper.java]: I removed them, but I can move this to a different PR.

Comment (rjeberhard) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/CrdHelper.java]: Long ago, we thought we would use these labels to know when to update the CRD... Now, we use the CRD's version field, so we don't need these labels.  They just cause confusion.

File: operator/src/main/java/oracle/kubernetes/operator/helpers/JobStepContext.java (hunk 9 above)
Comment (rjeberhard) [reply operator/src/main/java/oracle/kubernetes/operator/helpers/JobStepContext.java]: This is a surprise.  I thought the CheckStyle rule expected braces.

File: operator/src/main/resources/scripts/startNodeManager.sh (hunk 11 above)
Comment (davidzl) [reply operator/src/main/resources/scripts/startNodeManager.sh]: LOG_HOME is also provided in the create-domain-inputs.yaml:

# The in-pod location for domain log, server logs, server out, and node manager log files
# If not specified, the value is derived from the domainUID as /shared/logs/<domainUID>
logHome: /shared/logs/domain1

So I think we need include istio values in it too:

# experimental
# istioEnabled specifies whether or not it is deployed under Istio.
istioEnabled: false

# experimental
# istioReadinessPort specifies the WebLogic port for readiness probe which is used under Istio.
istioReadinessPort: 8888
